dex/networks/dcr/blockcache.go

```diff
@@ -0,0 +1,146 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package dcr
+
+import (
+	"fmt"
+	"sync"
+
+	"decred.org/dcrdex/dex"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	chainjson "github.com/decred/dcrd/rpc/jsonrpc/types/v2"
+)
+
+// Block defines basic information about a block.
+type Block struct {
+	Hash         *chainhash.Hash
+	Height       int64
+	PreviousHash string
+	Vote         bool // stakeholder vote result for the previous block
+}
+
+// BlockCache caches block information to prevent repeated calls to
+// rpcclient.GetblockVerbose.
+type BlockCache struct {
+	mtx       sync.RWMutex
+	blocks    map[chainhash.Hash]*Block
+	mainchain map[int64]*chainhash.Hash
+	bestBlock struct {
+		hash   *chainhash.Hash
+		height int64
+	}
+	log dex.Logger
+}
+
+func NewBlockCache(log dex.Logger) *BlockCache {
+	return &BlockCache{
+		blocks:    make(map[chainhash.Hash]*Block),
+		mainchain: make(map[int64]*chainhash.Hash),
+		log:       log,
+	}
+}
+
+// Add adds a block to the BlockCache. This method will translate the RPC result
+// to a *Block, returning the *Block. If the block is not orphaned, it will be
+// added to the mainchain.
+func (cache *BlockCache) Add(block *chainjson.GetBlockVerboseResult) (*Block, error) {
+	cache.mtx.Lock()
+	defer cache.mtx.Unlock()
+
+	hash, err := chainhash.NewHashFromStr(block.Hash)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding block hash %s: %w", block.Hash, err)
+	}
+
+	blk := &Block{
+		Hash:         hash,
+		Height:       block.Height,
+		PreviousHash: block.PreviousHash,
+		Vote:         block.VoteBits&1 != 0,
+	}
+	cache.blocks[*hash] = blk
```

Comment:
Given there's no obvious way to know when a certain block should be evicted, even for the caller, we may need to consider a capacity and some kind of eviction priority based on height.  But I don't think we have to do this right away though since these are small structures. A TODO comment would be fine.